airbyte-integrations/connectors/source-shopify/source_shopify/streams/base_streams.py

```diff
@@ -0,0 +1,677 @@
+            if stream_state:
+                params[self.filter_field] = stream_state.get(self.cursor_field)
+        return params
+
+    # Parse the `stream_slice` with respect to `stream_state` for `Incremental refresh`
+    # cases where we slice the stream, the endpoints for those classes don't accept any other filtering,
+    # but they provide us with the updated_at field in most cases, so we used that as incremental filtering during the order slicing.
+    def filter_records_newer_than_state(self, stream_state: Mapping[str, Any] = None, records_slice: Iterable[Mapping] = None) -> Iterable:
+        # Getting records >= state
+        if stream_state:
+            state_value = stream_state.get(self.cursor_field)
+            for record in records_slice:
+                if self.cursor_field in record:
+                    record_value = record.get(self.cursor_field, self.default_state_comparison_value)
+                    if record_value:
+                        if record_value >= state_value:
+                            yield record
+                    else:
+                        # old entities could have cursor field in place, but set to null
+                        self.logger.warning(
+                            f"Stream `{self.name}`, Record ID: `{record.get(self.primary_key)}` cursor value is: {record_value}, record is emitted without state comparison"
+                        )
+                        yield record
+                else:
+                    # old entities could miss the cursor field
+                    self.logger.warning(
+                        f"Stream `{self.name}`, Record ID: `{record.get(self.primary_key)}` missing cursor field: {self.cursor_field}, record is emitted without state comparison"
+                    )
+                    yield record
+        else:
+            yield from records_slice
+
+
+class IncrementalShopifySubstream(IncrementalShopifyStream):
+    """
+    IncrementalShopifySubstream - provides slicing functionality for streams using parts of data from parent stream.
+    For example:
+       - `Refunds Orders` is the entity of `Orders`,
+       - `OrdersRisks` is the entity of `Orders`,
+       - `DiscountCodes` is the entity of `PriceRules`, etc.
+
+    ::  @ parent_stream - defines the parent stream object to read from
+    ::  @ slice_key - defines the name of the property in stream slices dict.
+    ::  @ nested_record - the name of the field inside of parent stream record. Default is `id`.
+    ::  @ nested_record_field_name - the name of the field inside of nested_record.
+    ::  @ nested_substream - the name of the nested entity inside of parent stream, helps to reduce the number of
+          API Calls, if present, see `OrderRefunds` stream for more.
+    """
+
+    parent_stream_class: object = None
+    slice_key: str = None
+    nested_record: str = "id"
+    nested_record_field_name: str = None
+    nested_substream = None
```

Comment:
why does IncrementalShopifySubstream also have a nested_substream field? When should one use this class vs IncrementalShopifyNestedSubstream?

Reply:
Corrected the `nested_substream` > `nested_entity` for `IncrementalShopifyNestedSubstream`.  Renamed `IncrementalShopifyNestedSubstream` > `IncrementalShopifyNestedStream`.  To get rid of the confusion.